### PR TITLE
fix: handle disabled link tags, abort `adjustAmount` on change

### PR DESCRIPTION
**Pull request by sidvishnoi**
## Context

- Closes https://github.com/interledger/web-monetization-extension/issues/449
- And addresses some more bugs

## Changes proposed in this pull request

- Handle `disabled` link tags
- More:
   - If there's another call to `PaymentSession.adjustAmount`, abort existing ones.
   - Wait correctly for overpaying (handle more timing issues)
   - Some refactors

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">539f7460d1e78f9ccc9a6dd9298e92586430d773</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/10145964914">Run #10145964914</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/26546442324/artifacts/1751114687">Download</a></td></tr><tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/26546442324/artifacts/1751114693">Download</a></td></tr>
</tbody>
</table>

File: src/background/services/monetization.ts [hunk 1]
```diff
@@ -277,10 +298,25 @@ export class MonetizationService {
...
+      const currentTab = await getCurrentActiveTab(this.browser)
+      if (currentTab?.id) {
+        const idx = tabIds.indexOf(currentTab.id)
+        const tmp = tabIds[0]
+        tabIds[0] = currentTab.id
+        tabIds[idx] = tmp
+      }
+
+      for (const tabId of tabIds) {
+        const sessions = [...this.tabState.getSessions(tabId).values()]
```
Comment (sidvishnoi) [inline src/background/services/monetization.ts:315]: Should get enabled sessions only here too

File: src/background/services/monetization.ts [hunk 2]
```diff
@@ -354,3 +390,24 @@ export class MonetizationService {
...
+- A MONETIZATION LINK GETS DISABLED
+  * we send the `STOP_MONETIZATION` event for the disabled monetization link with the `intent` set to `disable` and we readjust the amount
+
+- A MONETIZATION LINK GETS ENABLED
+  * we send the `START_MONETIZATION` event and replace the previous session with a new one with the same requestId and we readjust the amount
+  !! we send the `RESUME_MONETIZATION` event for the enabled tag
+
+- THE HREF ATTRIBUTE IS UPDATED
+  * we send the `STOP_MONETIZATION` event for the updated monetization link with the `intent` set to `remove` and we send the `START_MONETIZATION` event for the new monetization link
+**/
```
Comment (sidvishnoi) [inline src/background/services/monetization.ts:413]:
```suggestion
```

File: src/background/services/monetization.ts [hunk 3]
```diff
@@ -277,10 +297,26 @@ export class MonetizationService {
...
+    this.events.on('storage.rate_of_pay_update', async ({ rate }) => {
       this.logger.debug("Received event='storage.rate_of_pay_update'")
-      for (const session of this.tabState.getAllSessions()) {
-        session.adjustAmount(rate)
+      const tabIds = this.tabState.getAllTabs()
+
+      // Move the current active tab to the front of the array
+      const currentTab = await getCurrentActiveTab(this.browser)
+      if (currentTab?.id) {
+        const idx = tabIds.indexOf(currentTab.id)
```
Comment (sidvishnoi) [inline src/background/services/monetization.ts:307]: Just to be sure, let's handle case if `idx === -1`

File: src/content/services/monetizationTagManager.ts [hunk 4]
```diff
@@ -180,9 +180,7 @@ export class MonetizationTagManager extends EventEmitter {
 
         handledTags.add(target)
       } else if (hasTarget && !typeSpecified) {
-        const stopMonetizationTag = this.onRemovedTag(target, true)
-        stopMonetizationTags.push(stopMonetizationTag)
-
+        stopMonetizationTags.push(this.onRemovedTag(target))
```
Comment (sidvishnoi) [inline src/content/services/monetizationTagManager.ts:183]: Just for a clearer diff:
```suggestion
        const stopMonetizationTag = this.onRemovedTag(target)
        stopMonetizationTags.push(stopMonetizationTag)

```

File: src/content/services/monetizationTagManager.ts [hunk 5]
```diff
@@ -197,12 +195,22 @@ export class MonetizationTagManager extends EventEmitter {
...
+              const { requestId, walletAddress } = this.getTagDetails(
+                target,
+                'onChangeDisabled'
+              )
+              if (isDisabled) {
+                stopMonetizationTags.push({ requestId, intent: 'disable' })
+              } else if (walletAddress) {
+                startMonetizationTags.push({ requestId, walletAddress })
+              }
+            } catch (error) {
```
Comment (sidvishnoi) [inline src/content/services/monetizationTagManager.ts:208]:
```suggestion
            } catch {
```

**Pull request merged**